File: 155-min-stack/min-stack.cpp
class MinStack {
public:
    vector <pair<int,int>>st;
    MinStack() {
        

        

        
    }
    
    void push(int val) {
        if(st.empty())
        {
        // first element so min is it
        pair <int,int> p ;
        p.first=val;
        p.second=val;
        st.push_back(p);
        }
        else
        {
           pair <int,int> p ;
           p.first = val;
           int pmin= st.back().second;
           p.second=min(pmin,val);
           st.push_back(p);

          
          
            
        }
    }

        
    
    
    void pop() {

        st.pop_back();

    }
    
    int top() {
        if(! st.empty())
        {
        pair <int,int> right_most = st.back();
        return right_most.first;
        }
        return -1;
        
        
        
    }
    
    int getMin() {
        pair <int,int> right_most = st.back();
        return right_most.second;
        
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 155-min-stack/min-stack.py
class MinStack:
   

    def __init__(self):
        self.stack=[]
        

    def push(self, val: int) -> None:
        
        if len(self.stack)==0:
            self.stack.append((val,val))
        else:
            last= self.stack[-1]

            lmin= min(last[1],val)

            self.stack.append((val,lmin))





        

    def pop(self) -> None:
        if len(self.stack) >0:
            self.stack.pop()


        

    def top(self) -> int:
        return self.stack[-1][0]
        

    def getMin(self) -> int:
        if len(self.stack)==0:
            return
        else:
            return self.stack[-1][-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()